### PR TITLE
Work In Progress – OpenTracing Brave Bridge

### PR DESCRIPTION
**Pull request by michaelsembwever**
Early code building an OpenTracing Brave bridge.

There's testing and tests still to go, and some of the new things in Brave-3 I'm not entirely confident of yet, but I do believe this is along the right track.

All spans are created with the local tracer. CS+CR and SR+SS annotations are done around the inject and join methods. Not sure yet whether that approach will work.

Haven't started on the async side of affairs either.

@adriancole @kristofa

### CODE REVIEW

File: opentracing-brave/pom.xml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>default-compile</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>compile</goal>
+                        </goals>
+                        <configuration>
+                            <source>1.8</source>
```
Comment (adriancole) [inline opentracing-brave/pom.xml]: :) this doesn't seem to agree with the comment

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            Optional<Span> parent,
+            Instant start,
+            Optional<ServerTracer> serverTracer) {
+
+        super(operationName, parent, start);
+        this.brave = brave;
+        this.serverTracer = serverTracer;
+
+        this.spanId = brave.localTracer().startNewSpan(
+                "???",
```
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: empty string is the latest unknown :P

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+final class BraveSpanImpl extends AbstractSpan {
+
+    final SpanId spanId;
+    final Optional<ServerTracer> serverTracer;
+    private final Brave brave;
+    private Optional<ClientTracer> clientTracer = Optional.empty();
+
+    public BraveSpanImpl(
+            Brave brave,
```
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: unused?

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java (hunk 2 above)
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: oh wait.. not sure which field this is.. if component, then binary annotation "lc"'s value is the component name

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    private final Brave brave = new Brave.Builder().build();
+
+    @Override
+    BraveSpanBuilderImpl createSpanBuilder() {
+        return new BraveSpanBuilderImpl(brave);
+    }
+
+    @Override
+    Map<String, String> getTraceState(Span span) {
+        return new HashMap<String,String>() {{
```
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java]: there's actually 4 possible fields to propagate, including parentId. the sampled flag is an interesting one here

https://github.com/openzipkin/brave/blob/master/brave-http/src/main/java/com/github/kristofa/brave/http/BraveHttpHeaders.java

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    Map<String, String> getBaggage(Span span) {
+        return ((BraveSpanImpl)span).baggage;
+    }
+
+    @Override
+    public <T> void inject(Span span, T carrier) {
+        brave.clientTracer().setClientSent();
```
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java]: interesting.. I didn't consider that inject had side-effects on the span itself.

**Comment by bensigelman** [comment]
@michaelsembwever @adriancole @kristofa just wanted to mention that I have great interest in this PR but know far less about Brave than any of you. That said, please pull me in about the OpenTracing side if it would be helpful.

File: opentracing-brave/pom.xml (hunk 1 above)
Comment (michaelsembwever) [inline opentracing-brave/pom.xml]: fixed

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java (hunk 2 above)
Comment (michaelsembwever) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: sry, what should the component value be?

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java (hunk 3 above)
Comment (michaelsembwever) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: fixed. forgot to call `localTracer.finish()` in `finish()`

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java (hunk 2 above)
Comment (adriancole) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: https://github.com/openzipkin/zipkin-java/blob/master/zipkin/src/main/java/zipkin/Constants.java#L138

ex "nginx" There's no implicit component in brave's local tracer. local
tracer was made for explicit use where the caller knows the component they
are invoking. maybe leave local tracer out, if there's no argument for
component we can hijack?

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java (hunk 4 above)
Comment (michaelsembwever) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java]: nice. used BraveHttpHeaders in a few places.

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java (hunk 5 above)
Comment (michaelsembwever) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveTracerImpl.java]: Yeah. I've stuck with using `Brave.localTracer()` by default, and hooking into `Brave.clientTracer()` in `inject(..)` and `Brave.serverTracer()` in `join(..)`.

I'm not sure if this approach works, it's just the simplest solution that comes to mind.

File: opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java (hunk 2 above)
Comment (michaelsembwever) [inline opentracing-brave/src/main/java/io/opentracing/propagation/BraveSpanImpl.java]: am just using "jvm" for now.

**Comment by kristofa** [comment]
@michaelsembwever Nice, but I'll have to do some reading on the OpenTracing api :)

**Comment by michaelsembwever** [comment]
re-opening this fresh as the destination is now master

**Pull request closed without merging**